File: python/day5.py
import re

def nice1(string):
    if sum([string.count(v) for v in 'aeiou']) < 3:
        return False
    if re.search("ab|cd|pq|xy", string) is not None:
        return False
    if re.search("(?P<r>.)(?P=r)", string):
        return True
    return False

def nice2(string):
    if re.search("(?P<r>..).*(?P=r)", string) is None:
        return False
    if re.search("(?P<r>.).(?P=r)", string) is None:
        return False
    return True

count1 = 0
count2 = 0
with open("day5.in") as fin:
    for line in fin:
        line = line.strip()

        if nice1(line):
            count1 += 1
        if nice2(line):
            count2 += 1
print(count1)
print(count2)

File: src/bin/day5.rs
extern crate pcre;

use pcre::Pcre;
use std::fs::File;
use std::io::BufReader;
use std::io::BufRead;

fn nice1(word: &String) -> bool {
    let mut one = Pcre::compile(r"(?:[aeiou].*?){3}").unwrap();
    let mut two = Pcre::compile(r"(.)\1").unwrap();
    let mut three = Pcre::compile(r"ab|cd|pq|xy").unwrap();

    if !one.exec(word).is_some() {
        return false;
    }
    if !two.exec(word).is_some() {
        return false;
    }
    if three.exec(word).is_some() {
        return false;
    }
    true
}

fn nice2(word: &String) -> bool {
    let mut one = Pcre::compile(r"(..).*?\1").unwrap();
    let mut two = Pcre::compile(r"(.).\1").unwrap();

    if !one.exec(word).is_some() {
        return false;
    }
    if !two.exec(word).is_some() {
        return false;
    }

    true
}

fn main() {
    let f = File::open("day5.in")
        .ok()
        .expect("Error opening input");
    let file = BufReader::new(&f);

    let words: Vec<String> = file.lines().map(|a| a.unwrap()).collect();

    println!("{}", words.clone().into_iter().filter(nice1).count());
    println!("{}", words.clone().into_iter().filter(nice2).count());
}
